### PR TITLE
feat: HMAC verification on IPN callback for security

### PR DESCRIPTION
**Pull request by Benjamin-Freoua-Alma**
### Reason for change

[Linear task](https://linear.app/almapay/issue/ECOM-1818/[-ps]-add-hmac-verification-on-ipn)

### Code changes

We add the call function to `isHmacValidated` in php client to check is the IPN callback is signed by Alma before to validate the order by IPN

### How to test

_As a reviewer, you are encouraged to test the PR locally._

Make an order, close the page before the return validation payment
Call the ipn callback with some other browser or API Client to check if the IPN need the Alma signature

### Checklist for authors and reviewers

- [ ] The title of the PR uses business wording, not technical jargon, for the changelog readers to understand it
- [ ] The PR implements the changes asked in the referenced task / issue
- [ ] The automated tests are compliant with the [testing strategy](https://www.notion.so/almapay/Backend-testing-strategy-06c642cec1bf47b9b8feca3a91ea8d4a)
- [ ] The tests are relevant, and cover the corner/error cases, not only the happy path
- [ ] You understand the impact of this PR on existing code/features
- [ ] The changes include adequate logging and Datadog traces
- [ ] Documentation is updated (API, developer documentation, ADR, Notion...)

### CODE REVIEW

File: alma/lib/Validators/PaymentValidation.php [hunk 1]
```diff
@@ -312,6 +322,9 @@ public function validatePayment($almaPaymentId)
      * @param $cartId
      *
      * @return \OrderCore|null
+     *
+     * @throws \PrestaShopDatabaseException
+     * @throws \PrestaShopException
      */
     private function getOrderByCartId($cartId)
```
Comment (Francois-Gomis) [inline alma/lib/Validators/PaymentValidation.php:329]: New throws in doc bloc , no try catch in  usage.

File: alma/lib/Validators/PaymentValidation.php [hunk 2]
```diff
@@ -347,4 +360,28 @@ private function getCartTotals($cart, $customer)
 
         return $cartTotals;
     }
+
+    /**
+     * @param $paymentId
```
Comment (Francois-Gomis) [inline alma/lib/Validators/PaymentValidation.php:365]: need params type

File: alma/lib/Validators/PaymentValidation.php [hunk 3]
```diff
@@ -347,4 +360,28 @@ private function getCartTotals($cart, $customer)
...
+            throw new PaymentValidationException(null, '[Alma] Payment ID is missing');
+        }
+        if (!$signature) {
+            throw new PaymentValidationException(null, '[Alma] Signature is missing');
+        }
+        if (!$this->paymentValidator->isHmacValidated($paymentId, $apiKey, $signature)) {
+            throw new PaymentValidationException(null, '[Alma] Signature is invalid');
+        }
+
+        return true;
```
Comment (Francois-Gomis) [inline alma/lib/Validators/PaymentValidation.php:385]: Return not needed

File: alma/lib/Validators/PaymentValidation.php [hunk 4]
```diff
@@ -347,4 +360,28 @@ private function getCartTotals($cart, $customer)
 
         return $cartTotals;
     }
+
+    /**
+     * @param $paymentId
+     * @param $apiKey
+     * @param $signature
+     *
+     * @return true
```
Comment (Francois-Gomis) [inline alma/lib/Validators/PaymentValidation.php:369]: no return needed

File: alma/tests/Unit/Validators/PaymentValidationTest.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            $this->clientPaymentValidator
+        );
+    }
+
+    public function tearDown()
+    {
+        $this->paymentValidation = null;
+    }
+
+    /**
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Validators/PaymentValidationTest.php:63]: Need a test without Api Key , you can define a dataprovider with bad combinations

File: alma/tests/Unit/Validators/PaymentValidationTest.php [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        $this->clientPaymentValidator->expects($this->once())
+            ->method('isHmacValidated')
+            ->with(self::PAYMENT_ID, self::API_KEY, self::WRONG_SIGNATURE)
+            ->willReturn(false);
+        $this->expectException(PaymentValidationException::class);
+        $this->paymentValidation->checkSignature(self::PAYMENT_ID, self::API_KEY, self::WRONG_SIGNATURE);
+    }
+
+    /**
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Validators/PaymentValidationTest.php:94]: no throws on this test - if you remove true in method, juste test no exception thrown

File: alma/lib/Validators/PaymentValidation.php [hunk 7]
```diff
@@ -347,4 +360,28 @@ private function getCartTotals($cart, $customer)
...
+     *
+     * @throws PaymentValidationException
+     */
+    public function checkSignature($paymentId, $apiKey, $signature)
+    {
+        if (!$paymentId) {
```
Comment (Francois-Gomis) [inline alma/lib/Validators/PaymentValidation.php:375]: i think this test is necessary only for log, isHmacValidated test if params are string

**Review by Francois-Gomis — CHANGES_REQUESTED** [review]

File: alma/lib/Traits/BuilderTrait.php [hunk 8]
```diff
@@ -1202,4 +1203,18 @@ public function getInsuranceApiService($insuranceApiService = null)
...
     }
+
+    /**
+     * @param PaymentValidator $clientPaymentValidator
+     *
+     * @return PaymentValidator
+     */
+    public function getClientPaymentValidator($clientPaymentValidator = null)
+    {
+        if ($clientPaymentValidator) {
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Traits/BuilderTrait.php:1214]: @Francois-Gomis What do you mean ? Remove the condition if ?
I don't completely agree with this builder yet.

File: alma/lib/Validators/PaymentValidation.php (hunk 1 above)
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Validators/PaymentValidation.php:329]: @Francois-Gomis What do you mean ? Remove the doc bloc about throws ?

File: alma/lib/Validators/PaymentValidation.php (hunk 7 above)
Comment (Benjamin-Freoua-Alma) [reply alma/lib/Validators/PaymentValidation.php:375]: @Francois-Gomis Do you want I replace throw by log ?

File: alma/controllers/front/ipn.php [hunk 9]
```diff
@@ -42,6 +43,11 @@ class AlmaIpnModuleFrontController extends ModuleFrontController
 
     public $ssl = true;
 
+    /**
+     * @var PaymentValidation
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/controllers/front/ipn.php:55]: @Francois-Gomis is removed by lint, not me :/

File: alma/exceptions/PaymentValidationException.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Alma\PrestaShop\Exceptions;
+
+if (!defined('_PS_VERSION_')) {
+    exit;
+}
+
+class PaymentValidationException extends AlmaException
+{
+    public $cart;
```
[the remark this answers is not in the document]
Comment (Benjamin-Freoua-Alma) [reply alma/exceptions/PaymentValidationException.php:36]: @Francois-Gomis is removed by lint, not me :/

File: alma/lib/Validators/PaymentValidation.php (hunk 7 above)
Comment (Francois-Gomis) [reply alma/lib/Validators/PaymentValidation.php:373]: No, need, if an error occur, you have exception log. If no error no logs

File: alma/tests/Unit/Validators/PaymentValidationTest.php [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+            ->method('isHmacValidated')
+            ->with(self::PAYMENT_ID, self::API_KEY, self::GOOD_SIGNATURE)
+            ->willReturn(true);
+        $this->paymentValidation->checkSignature(self::PAYMENT_ID, self::API_KEY, self::GOOD_SIGNATURE);
+    }
+
+    /**
+     * @return array[]
+     */
+    public function checkSignatureWrongParamsDataProvider()
```
Comment (Francois-Gomis) [inline alma/tests/Unit/Validators/PaymentValidationTest.php:102]: Are you sure, api is always a string '' not null ?

File: alma/tests/Unit/Validators/PaymentValidationTest.php (hunk 6 above)
Comment (Francois-Gomis) [reply alma/tests/Unit/Validators/PaymentValidationTest.php:87]: In your comment you have a throws decorator

**Review by Francois-Gomis — APPROVED** [review]

File: alma/tests/Unit/Validators/PaymentValidationTest.php (hunk 11 above)
Comment (Benjamin-Freoua-Alma) [reply alma/tests/Unit/Validators/PaymentValidationTest.php:102]: @Francois-Gomis I added the test with null

**Review by joyet-simon — APPROVED** [review]

**Pull request merged**